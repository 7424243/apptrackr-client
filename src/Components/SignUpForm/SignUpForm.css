/* Classes */

.password_reqs {
  line-height: 1;
  overflow-wrap: break-word;
}

.signup_container {
    display: flex;
    flex-direction: column;
    font-size: larger;
}

.signup_error {
  margin-left: 20px;
  margin-right: 20px;
  text-align: left;
  color: #e27802;
  font-weight: bold;
}

.signup_form {
    margin-left: 20px;
    padding: 40px 0;
    max-width: 80%;
    text-align: left;
}

.signup_input {
  font-size: large;
  border-style: solid;
}

.signup_section {
    margin-bottom: 20px;
}

/*Spinner*/

.lds-ripple {
  display: inline-block;
  position: relative;
  width: 80px;
  height: 80px;
}

.lds-ripple div {
  position: absolute;
  border: 4px solid #686d76;
  opacity: 1;
  border-radius: 50%;
  animation: lds-ripple 1s cubic-bezier(0, 0.2, 0.8, 1) infinite;
}

.lds-ripple div:nth-child(2) {
  animation-delay: -0.5s;
}

@keyframes lds-ripple {
  0% {
    top: 36px;
    left: 36px;
    width: 0;
    height: 0;
    opacity: 1;
  }
  100% {
    top: 0px;
    left: 0px;
    width: 72px;
    height: 72px;
    opacity: 0;
  }
}

/* Media Query */

@media only screen and (min-width: 650px) {

  .signup_error {
    text-align: center;
  }  
  
  .signup_form {
        margin: 0 auto;
    }

}

