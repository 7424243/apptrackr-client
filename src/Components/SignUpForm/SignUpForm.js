import React, {Component} from 'react'
import ApptrackrContext from '../../ApptrackrContext'
import config from '../../config'
import TokenService from '../../services/token-service'
import ErrorBoundary from '../ErrorBoundary/ErrorBoundary'
import RecButton from '../RecButton/RecButton'
import ValidationError from '../ValidationError/ValidationError'
import './SignUpForm.css'

class SignUpForm extends Component {

    state = {
        full_name: '',
        user_name: '',
        password: '',
        error: null
    }

    static contextType = ApptrackrContext

    handleSubmit = e => {
        e.preventDefault()
        fetch(`${config.API_ENDPOINT}/users/`, {
            method: 'POST',
            headers: {
                'content-type': 'application/json',
            },
            body: JSON.stringify(this.state)
        })
            .then(res => {
                if(!res.ok) {
                    return res.json().then(e => Promise.reject(e))
                }
                return res.json()
            })
            .then(() => {
                const {user_name, password} = this.state
                const loginCreds = {user_name, password}
                //automatically login after signing up
                return fetch(`${config.API_ENDPOINT}/auth/login`, {
                    method: 'POST',
                    headers: {
                        'content-type': 'application/json',
                    },
                    body: JSON.stringify(loginCreds)
                })
                    .then(res => {
                        if(!res.ok) {
                            return res.json().then(e => Promise.reject(e))
                        }
                        return res.json()
                    })
                    .then(res => {
                        TokenService.saveAuthToken(res.authToken)
                        this.context.onLoginSuccess()
                        this.props.history.push('/jobapps')
                    })
                    .catch(err => {
                        this.setState({error: err.error})
                        console.error({err})
                    })
            })
            .catch(err => {
                this.setState({error: err.error.message})
                console.error({err})
            })
    }

    handleAddFullName = e => {
        this.setState({full_name: e.target.value})
    }

    handleAddUserName = e => {
        this.setState({user_name: e.target.value})
    }

    handleAddPassword = e => {
        this.setState({password: e.target.value})
    }

    validateFullName() {
        const fullName = this.state.full_name.trim()
        if(fullName.length === 0) {
            return 'A full name is required'
        }
    }

    validateUserName() {
        const userName = this.state.user_name.trim()
        if(userName.length === 0) {
            return 'A username is required'
        }
    }

    validatePassword() {
        //eslint-disable-next-line
        const REGEX_UPPER_LOWER_NUMBER_SPECIAL = /(?=.*[a-z])(?=.*[A-Z])(?=.*[0-9])(?=.*[!@#\$%\^&])\S]+/
        const password = this.state.password.trim()
        if(password.length < 8) {
            return 'Password must be longer than 8 characters'
        }
        if(password.length > 72) {
            return 'Password must be less than 72 characters'
        }
        if(password.startsWith(' ') || password.endsWith(' ')) {
            return 'Password must not start or end with empty spaces'
        }
        if(!REGEX_UPPER_LOWER_NUMBER_SPECIAL.test(password)) {
            return 'Password must contain at least 1 upper case, 1 lower case, 1 number and 1 special character'
        }
    }

    render() {
        return (
            <div className='signup_container'>
                <ErrorBoundary>
                    <form 
                        className='signup_form'
                        onSubmit={this.handleSubmit}
                    >
                        <h2>Sign Up Form</h2>
                        <p>* Required</p>
                        <section className='signup_input'>*
                            <input 
                                type='text' 
                                placeholder='Full Name'
                                autoComplete='off'
                                required
                                onChange={this.handleAddFullName}
                            />
                            {this.state.full_name && <ValidationError message={this.validateFullName()}/>}
                        </section>
                        <section className='signup_input'>*
                            <input 
                                type='text'
                                placeholder='Username'
                                autoComplete='off'
                                required
                                onChange={this.handleAddUserName}
                            />
                            {this.state.user_name && <ValidationError message={this.validateUserName()}/>}                        
                        </section>
                        <section className='signup_input'>
                            <p>* Must be {'>'} 8 characters and contain at least 1 upper case, 1 lower case, 1 number, and 1 special character.</p>
                            <input 
                                type='password' 
                                placeholder='Password'
                                autoComplete='off'
                                required
                                onChange={this.handleAddPassword}
                            />      
                            {this.state.password && <ValidationError message={this.validatePassword()}/>}                  
                        </section>
                            <RecButton type='submit'>Sign Up</RecButton>
                    </form>
                    {this.state.error && <p>{this.state.error}</p>}
                </ErrorBoundary>
                
            </div>
        )
    }
}

export default SignUpForm