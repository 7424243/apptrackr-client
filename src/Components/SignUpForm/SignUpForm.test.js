import React from 'react'
import ReactDOM from 'react-dom'
import renderer from 'react-test-renderer'
import {BrowserRouter} from 'react-router-dom'
import SignUpForm from './SignUpForm'

describe('SignUpForm Component', () => {

    //smoke test
    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDOM.render(<BrowserRouter><SignUpForm/></BrowserRouter>, div)
        ReactDOM.unmountComponentAtNode(div)
    })

    //snapshot test
    it('renders the UI as expected', () => {
        const tree = renderer
            .create(<BrowserRouter><SignUpForm/></BrowserRouter>)
            .toJSON()
        expect(tree).toMatchSnapshot()
    })
    
})