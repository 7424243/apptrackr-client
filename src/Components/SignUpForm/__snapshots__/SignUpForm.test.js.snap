// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignUpForm Component renders the UI as expected 1`] = `
<div
  className="signup_container"
>
  <form
    className="signup_form"
    onSubmit={[Function]}
  >
    <h2>
      Sign Up Form
    </h2>
    <p>
      * Required
    </p>
    <section
      className="signup_input"
    >
      *
      <input
        autoComplete="off"
        onChange={[Function]}
        placeholder="Full Name"
        required={true}
        type="text"
      />
      
    </section>
    <section
      className="signup_input"
    >
      *
      <input
        autoComplete="off"
        onChange={[Function]}
        placeholder="Username"
        required={true}
        type="text"
      />
      
    </section>
    <section
      className="signup_input"
    >
      <p>
        * Must be 
        &gt;
         8 characters and contain at least 1 upper case, 1 lower case, 1 number, and 1 special character.
      </p>
      <input
        autoComplete="off"
        onChange={[Function]}
        placeholder="Password"
        required={true}
        type="password"
      />
      
    </section>
    <div
      className="recbutton_content"
    >
      <button
        className="recbutton"
        type="submit"
      >
        Sign Up
      </button>
    </div>
  </form>
</div>
`;
